pkg/doc/jose/jwk.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package jose
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/big"
+	"strings"
+
+	"github.com/decred/dcrd/dcrec/secp256k1/v2"
+	"github.com/square/go-jose/v3"
+	"golang.org/x/crypto/ed25519"
+)
+
+const (
+	secp256k1Crv = "secp256k1"
+	secp256k1Kty = "EC"
+	bitsPerByte  = 8
+)
+
+// JWK (JSON Web Key) is a JSON data structure that represents a cryptographic key.
+type JWK jose.JSONWebKey
+
+// JWK gets JWK from JOSE headers.
+func (h Headers) JWK() (*JWK, bool) {
+	jwkRaw, ok := h[HeaderJSONWebKey]
+	if !ok {
+		return nil, false
+	}
+
+	var jwk JWK
+
+	err := convertMapToValue(jwkRaw, &jwk)
+	if err != nil {
+		return nil, false
+	}
+
+	return &jwk, true
+}
+
+type jsonWebKey struct {
+	Kty string      `json:"kty,omitempty"`
+	Kid string      `json:"kid,omitempty"`
+	Crv string      `json:"crv,omitempty"`
+	X   *byteBuffer `json:"x,omitempty"`
+	Y   *byteBuffer `json:"y,omitempty"`
+}
+
+// DecodePublicKey reads a public key from its JSON Web Key representation.
+func DecodePublicKey(jwkBytes []byte) ([]byte, error) {
+	key := &jsonWebKey{}
+
+	err := json.Unmarshal(jwkBytes, key)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read JWK, %w", err)
+	}
+
+	if strings.EqualFold(key.Kty, secp256k1Kty) && strings.EqualFold(key.Crv, secp256k1Crv) {
```

Comment:
some comments on the various conditions in this function would be helpful (linked to the DID document data model).

Reply:
Are there other cases we aren't covering yet?

Reply:
secp256k1 is not the only EC key type that should be supported by this JWK implementation. We need to support NIST curves as well (secp256r1 or P-2516, secp384r1 or P-384 and secp521r1 or P-521), these are EC curves as well.

Reply:
To make it more generic, I would override `UnmarshalJSON()` of `JWK` to support `secp256k1` elliptic curve (`MarshallJSON()` for symmetry). And I would create e.g. `Bytes()` method returning `[]byte` for `JWK` as well where `x509.MarshalPKIXPublicKey()` could be used.

Reply:
@sudeshrshetty the above is just my personal opinion, I'm OK with your solution in general.

Reply:
@troyronda almost all cases are covered by go-jose, my implementation only covers 'EC/secp256k1', rest of the types will be handled by go-jose.

Reply:
@kdimak sounds good actually, I agree. I will add that as part of upcoming PR. Created an issue #1513  for this.